### DISCUSSION TITLE
How to update the stream only on needed events?

### DISCUSSION THREAD

**Discussion by ivanburlakov** [simolus3/drift#1391] [Q&A]
In my app I have a `Moor` `Stream` list of database items. I need to update a stream when new entry inserts or deletes from the table. But in my UI I also need a live changable title `TextField`, that updates the database entry everytime I type a letter (video below). The problem is, when I update the database entry with `moor` - the `Stream` update also triggers, rerenders the whole list and puts my cursor to the beginning. To solve this problem I connected another raw `sqlite3` database library to update the table without triggering the stream. It works ok, but when I change the query parameters of the `Stream` (for example I change month, like on the video below), it flickers for a moment with the prevoius (unedited) text (although when I open it again - it is edited without a problem). I think it's due to the fact I am using a `AnimatedSwitcher` and it is using the unupdated list items from `Stream` to draw it in `itemRemoveBuilder`.

The question is: what is the best way to solve this problem? Maybe `moor` has some methods, that I do not know about. Maybe I could not use another db library like `sqlite3` at all? Thank you!

https://user-images.githubusercontent.com/66262596/127741153-7f35b49e-68d8-4222-b75a-7eb220ba0334.mp4

**Comment by simolus3** [comment]
I think moor is behaving as intended here. You're writing to a database, therefore you'll be receiving stream updates. There's no way to disable this behavior.

You could [pause](https://api.dart.dev/stable/2.13.4/dart-async/StreamSubscription/pause.html) the stream when you open a text field and [resume](https://api.dart.dev/stable/2.13.4/dart-async/StreamSubscription/resume.html) it when you're done editing?

**Reply by ivanburlakov** [reply]
User can just navigate/close the app in the middle of editing, so it would add a lot of complexity to save in every case.

I think the right decision would be to take the text from the database only on first initialisation of element. But don't know how to do it, cause the code in `ListView.builder` reruns on every update of a database.

**Reply by ivanburlakov** [reply]
I think I solved it by creating a `List` of `id`s in my widget, giving an initial value to the `TextField`on render only if it's not in this `List` and adding the `id` in this `List` AFTER defining the `TextField`, for example:
```
{
final widget = ...
if (!list.contains(id)) list.add(id)
 return widget;
}
```
Also I clear the list on `dispose` and on every query change (switching month in my example)

I don't know if it is a reliable solution, so if someone knows better way - would be nice to know :)
